apps/passport-server/src/services/poapService.ts

```diff
@@ -0,0 +1,186 @@
+        throw new Error("proof must be ZKEdDSAEventTicketPCD type");
+      }
+
+      const pcd = await ZKEdDSAEventTicketPCDPackage.deserialize(parsed.pcd);
+
+      if (!process.env.SERVER_EDDSA_PRIVATE_KEY)
+        throw new Error(`missing server eddsa private key .env value`);
+
+      const TICKETING_PUBKEY = await getEdDSAPublicKey(
+        process.env.SERVER_EDDSA_PRIVATE_KEY
+      );
+
+      const signerMatch =
+        pcd.claim.signer[0] === TICKETING_PUBKEY[0] &&
+        pcd.claim.signer[1] === TICKETING_PUBKEY[1];
+
+      span?.setAttribute("signerMatch", signerMatch);
+
+      if (!signerMatch) {
+        throw new Error("signer of PCD is invalid");
+      }
+
+      if (!(await ZKEdDSAEventTicketPCDPackage.verify(pcd))) {
+        throw new Error("pcd invalid");
+      }
+
+      const {
+        validEventIds,
+        partialTicket: { ticketId }
+      } = pcd.claim;
+
+      if (ticketId == null) {
+        throw new Error("ticket ID must be revealed");
+      }
+      const devconnectPretixTicket =
+        await fetchDevconnectPretixTicketByTicketId(
+          this.context.dbPool,
+          ticketId
+        );
+      if (devconnectPretixTicket == null) {
+        throw new Error("ticket ID does not exist");
+      }
+      const { devconnect_pretix_items_info_id, is_consumed } =
+        devconnectPretixTicket;
+
+      span?.setAttribute("ticketId", ticketId);
+
+      if (
+        !(
+          validEventIds &&
+          validEventIds.length === 1 &&
+          validEventIds[0] === DEVCONNECT_COWORK_SPACE_EVENT_ID
+        )
+      ) {
+        throw new Error(
+          "valid event IDs of PCD does not match Devconnect Cowork space"
+        );
+      }
+
+      span?.setAttribute("isConsumed", is_consumed);
+
+      if (!is_consumed) {
+        throw new Error("ticket has not been consumed");
+      }
+
+      span?.setAttribute("productId", devconnect_pretix_items_info_id);
+
+      if (
+        !DEVCONNECT_COWORK_SPACE_VALID_PRODUCT_IDS.includes(
+          devconnect_pretix_items_info_id
+        )
+      ) {
+        throw new Error("product ID is invalid");
+      }
+
+      return ticketId;
+    });
+  }
+
+  public async getDevconnectPoapClaimUrl(
+    serializedPCD: string
+  ): Promise<string> {
+    try {
+      const ticketId = await this.validateDevconnectTicket(serializedPCD);
+
+      // We have already checked that `ticketId` is defined in validateDevconnectPCD
+      const hashedTicketId = await getHash(ticketId);
+      const existingPoapLink = await getExistingClaimUrlByTicketId(
+        this.context.dbPool,
+        hashedTicketId
+      );
+      if (existingPoapLink != null) {
+        return existingPoapLink;
+      }
+
+      const newPoapLink = await claimNewPoapUrl(
+        this.context.dbPool,
+        "devconnect",
+        hashedTicketId
+      );
+      if (newPoapLink == null) {
+        throw new Error("Not enough Devconnect POAP links");
+      }
```

Comment:
there is still a slight concurrency issue here that might let a single user claim multiple poap links say, if they send a burst of 1000 requests at once, some of them are bound to concurrently get to line 156. i'm not sure it's a huge concern. that being said, a fix may be quite simple. [this package](https://www.npmjs.com/package/async-lock) could let you surround this section with a mutex such that your two queries are guaranteed to run in isolation per-ticketId.